src/append/syslog.rs

```diff
@@ -0,0 +1,247 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Appender for writing log records to syslog.
+//!
+//! # Examples
+//!
+//!```rust, no_run
+//! use logforth::append::syslog;
+//! use logforth::append::syslog::Syslog;
+//! use logforth::append::syslog::SyslogWriter;
+//!
+//! let syslog_writer = SyslogWriter::tcp_well_known().unwrap();
+//! let (non_blocking, _guard) = syslog::non_blocking_builder().finish(syslog_writer);
+//!
+//! logforth::builder()
+//!     .dispatch(|d| {
+//!         d.filter(log::LevelFilter::Trace)
+//!             .append(Syslog::new(non_blocking))
+//!     })
+//!     .apply();
+//!
+//! log::info!("This log will be written to syslog.");
+//! ```
+
+use std::io;
+
+use fasyslog::sender::SyslogSender;
+use fasyslog::SDElement;
+use log::Record;
+
+use crate::non_blocking::NonBlocking;
+use crate::non_blocking::NonBlockingBuilder;
+use crate::non_blocking::Writer;
+use crate::Append;
+use crate::Layout;
+
+// re-exports to avoid version conflicts
+mod exported {
+    pub use fasyslog::format::SyslogContext;
+    pub use fasyslog::Facility;
+}
+pub use exported::*;
+
+/// The format of the syslog message.
+#[derive(Debug, Copy, Clone)]
+pub enum SyslogFormat {
+    /// [RFC 3614] (BSD syslog Protocol)
+    ///
+    /// [RFC 3164]: https://datatracker.ietf.org/doc/html/rfc3164
+    RFC3164,
+    /// [RFC 5424] (The Syslog Protocol)
+    ///
+    /// [RFC 5424]: https://datatracker.ietf.org/doc/html/rfc5424
+    RFC5424,
+}
+
+/// An appender that writes log records to syslog.
+#[derive(Debug)]
+pub struct Syslog {
+    writer: NonBlocking<SyslogWriter>,
+    format: SyslogFormat,
+    context: SyslogContext,
+    layout: Option<Layout>,
+}
+
+impl Syslog {
+    /// Creates a new [`Syslog`] appender.
+    pub fn new(writer: NonBlocking<SyslogWriter>) -> Self {
+        Self {
+            writer,
+            format: SyslogFormat::RFC3164,
+            context: SyslogContext::default(),
+            layout: None,
+        }
+    }
+
+    /// Set the format of the [`Syslog`] appender.
+    pub fn with_format(mut self, format: SyslogFormat) -> Self {
+        self.format = format;
+        self
+    }
+
+    /// Set the context of the [`Syslog`] appender.
+    pub fn with_context(mut self, context: SyslogContext) -> Self {
+        self.context = context;
+        self
+    }
+
+    /// Set the layout of the [`Syslog`] appender.
+    ///
+    /// Default to `None`, only the args will be logged.
+    pub fn with_layout(mut self, layout: impl Into<Layout>) -> Self {
```

Comment:
The builder style is not inline with `OpentelemetryLogBuilder`:

```
let builder = OpentelemetryLogBuilder::new("my_service", "http://localhost:4317");
builder.label("env", "production");
```

Maybe choose one and apply for all of them?

Reply:
@andylokandy You may see https://github.com/fast/logforth/pull/72/commits/41640f74937fba1690c784965fa123bbfa9feba8 and https://github.com/fast/logforth/pull/72/commits/0c8dce7d084c20dd671ce8a08caf4cb82e05563c.

I ever think of using one pattern, but later found that `layout` is good for builder, while other appenders are not builders, so `layout` can be misleading with a getter method.

Reply:
Or maybe anyway they are builder-alike so that we treat all of them builders 🤣

Reply:
I see. I'll prefer `with_layout`. A single word `.layout` sounds like a verb, while it actually has no side effect...

Reply:
ping

Reply:
Do you mean that we should change methods of `OpentelemetryLogBuilder` to align with the `with_` flavor?

Reply:
Others should stay with the current `with_` flavor and nothing need to be changed.